Packages/com.unity.inputsystem/InputSystem/Plugins/Users/InputUser.cs

```diff
@@ -666,43 +672,40 @@ public void UnpairDevices()
         {
             var userIndex = index; // Throws if user is invalid.
 
-            // Reset device count so it appears all devices are gone from the user. We still want to send
-            // notifications one by one, so we can't yet remove the devices from s_AllPairedDevices.
-            var deviceCount = s_AllUserData[userIndex].deviceCount;
-            var deviceStartIndex = s_AllUserData[userIndex].deviceStartIndex;
-            s_AllUserData[userIndex].deviceCount = 0;
-            s_AllUserData[userIndex].deviceStartIndex = 0;
+            RemoveLostDevicesForUser(userIndex);
 
-            // Update actions, if necessary.
-            var actions = s_AllUserData[userIndex].actions;
-            if (actions != null)
-                actions.devices = null;
+            using (InputActionRebindingExtensions.DeferBindingResolution())
+            {
+                // We could remove the devices in bulk here but we still have to notify one
+                // by one which ends up being more complicated than just unpairing the devices
+                // individually here.
+                while (s_AllUserData[userIndex].deviceCount > 0)
+                    UnpairDevice(s_AllPairedDevices[s_AllUserData[userIndex].deviceStartIndex + s_AllUserData[userIndex].deviceCount - 1]);
+            }
 
             // Update control scheme, if necessary.
             if (s_AllUserData[userIndex].controlScheme != null)
                 UpdateControlSchemeMatch(userIndex);
+        }
 
-            // Notify.
-            for (var i = 0; i < deviceCount; ++i)
-                Notify(userIndex, InputUserChange.DeviceUnpaired, s_AllPairedDevices[deviceStartIndex + i]);
-
-            // Remove.
-            ArrayHelpers.EraseSliceWithCapacity(ref s_AllPairedDevices, ref s_AllPairedDeviceCount, deviceStartIndex, deviceCount);
-            if (s_AllUserData[userIndex].lostDeviceCount > 0)
+        private static void RemoveLostDevicesForUser(int userIndex)
+        {
+            var lostDeviceCount = s_AllUserData[userIndex].lostDeviceCount;
+            if (lostDeviceCount > 0)
```

Comment:
>to reduce nesting

Wait, that *increases* nesting, right?

Personally, I prefer the opposite, i.e. code that isn't heavily indented. I.e. instead of

```
class Foo
{
    void Bar()
    {
        if (x)
        {
             Flub();
             if (y)
             {
                 Blub();
```

get

```
class Foo
{
    void Bar()
    {
        if (!x)
            return;

         Flub();
         if (!y)
             return;

         Blub();
```

Aside from avoid brace-craziness, every time you see a `return`, you know it's the exit right there. No need to jump all the way to the end of the `if` to see whether there's something else coming in after the block.

Reply:
LOL... so first I respond to the *opposite* of what you're saying and then I accidentally hit the "Resolve Conversation" button. Guess it's Friday...

So.... let me give this another try :)

Yup, fully agree with you there. `if` should be inverted here. Ignore my rambling :)

Reply:
All good :)